Drop in cfg files support

pmores opened the pull request:
This PR adds support for drop-in configuration files stored in a `config.d` directory at the same path where the main `configuration.toml` file is located.

The PR is meant to be backwards-compatible - if no drop-in directory is found or if it's empty, the current behaviour is preserved.  Drop-ins will only override the specific settings that they contain, there should be no collateral damage to other settings loaded from the main `configuration.toml` anywhere in the configuration.  Drop-in files handling aims to follow existing conventions - the files are evaluated in alphabetical order and the last one to modify a specific setting wins.

Individual commits in this series are ordered bottom-up so that earlier commits build up foundations that later commits then use until the new functionality is finally fully integrated into pre-existing code in the last commit.

fidencio commented:
@pmores , without going through the patches yet, mind to add some notes on how you're dealing with different sections of the default configuration?

Also, it'd be good to know, before going through the series, what's the expectation of precedence order and whatnot on files touching the very same configuration bits.

fidencio commented:
Last but not least, we need tests, we really need tests covering as much cases as possible and those should be part of this PR.

pmores commented:
> @pmores , without going through the patches yet, mind to add some notes on how you're dealing with different sections of the default configuration?

Hi, I'm not sure if I understand the question properly but there's no differential handling of default configuration sections.  Whatever can be configured in `configuration.toml` can also be modified by a drop-in file.
 
> Also, it'd be good to know, before going through the series, what's the expectation of precedence order and whatnot on files touching the very same configuration bits.

This PR means to follow the usual way drop-ins are handled. It enumerates the drop-in directory in an alphabetical order and the expectation is that users use the familiar naming conventions to ensure the desired order (calling the drop-ins `10-this`, `20-that` etc.).  The last file to change a specific setting wins.

pmores commented:
> Last but not least, we need tests, we really need tests covering as much cases as possible and those should be part of this PR.

I agree and I've looked into adding tests already but frankly I couldn't figure out so far how to do that.  So I thought I'd let people review the change itself while I'm putting more effort into formulating tests. :-)

fidencio commented:
> > @pmores , without going through the patches yet, mind to add some notes on how you're dealing with different sections of the default configuration?
> 
> Hi, I'm not sure if I understand the question properly but there's no differential handling of default configuration sections. Whatever can be configured in `configuration.toml` can also be modified by a drop-in file.

For instance, if I have a drop-in file with the [runtime] section, where I just change the vfio_mode.
Does it replace all the values of the current [runtime] section?  Or does it only override the ones that I explicitly mentioned (vfio_mode)?

> 
> > Also, it'd be good to know, before going through the series, what's the expectation of precedence order and whatnot on files touching the very same configuration bits.
> 
> This PR means to follow the usual way drop-ins are handled. It enumerates the drop-in directory in an alphabetical order and the expectation is that users use the familiar naming conventions to ensure the desired order (calling the drop-ins `10-this`, `20-that` etc.). The last file to change a specific setting wins.

Perfect!

pmores commented:
> For instance, if I have a drop-in file with the [runtime] section, where I just change the vfio_mode. Does it replace all the values of the current [runtime] section? Or does it only override the ones that I explicitly mentioned (vfio_mode)?

Oh, this PR goes to great lengths to ensure that just `vfio_mode` will be overridden by the drop-in in your scenario.

In fact, this is the default behaviour of BurntSushi _unless_ a struct that corresponds to a config file section is stored in a `map` (`Hypervisor` and `Agent`) in which case BurntSushi wipes out the previous contents of the struct that came for the main `configuration.toml`.  The bulk of this PR is actually about fixing this.  All this is described in more detail in the PR itself, both in commit messages and code comments.

fidencio commented:
> > For instance, if I have a drop-in file with the [runtime] section, where I just change the vfio_mode. Does it replace all the values of the current [runtime] section? Or does it only override the ones that I explicitly mentioned (vfio_mode)?
> 
> Oh, this PR goes to great lengths to ensure that just `vfio_mode` will be overridden by the drop-in in your scenario.
> 
> In fact, this is the default behaviour of BurntSushi _unless_ a struct that corresponds to a config file section is stored in a `map` (`Hypervisor` and `Agent`) in which case BurntSushi wipes out the previous contents of the struct that came for the main `configuration.toml`. The bulk of this PR is actually about fixing this. All this is described in more detail in the PR itself, both in commit messages and code comments.

Please, **also** add this as part of the description of the PR itself.  You can edit it here in GitHub.

I appreciate the effort on explaining everything well on the commit messages, I really do, but folks may want to have a grasp of what the PR is doing, what's the proposal and the approach taken, before having to go through the patches and actually review the PR, or to go through the thread of exchanged messages.

fidencio commented:
@pmores, I'm thinking about users using this one and I have a question for you (and also for the other folks who'll review this one).

Consider I'm using kata-deploy, and while tweaking things here in there I connected to the node and added my `config.d` folder and dropped in my files there.  What would be the expected kata-deploy behaviour from your POV?   Shall it remove the `config.d` directory entirely?  Shall it keep it there?

pmores commented:
> @pmores, I'm thinking about users using this one and I have a question for you (and also for the other folks who'll review this one).
> 
> Consider I'm using kata-deploy, and while tweaking things here in there I connected to the node and added my `config.d` folder and dropped in my files there. What would be the expected kata-deploy behaviour from your POV? Shall it remove the `config.d` directory entirely? Shall the `config.d` directory be kept there?

Admittedly I'm not familiar enough with the behaviour kata-deploy aims for to give an authoritative answer, but my initial instinct would probably be to preserve site-specific customisations. (?)

pmores commented:
Pushed a new version that
- adds tests
- includes a bunch of formatting fixes to placate `gofmt`.

pmores commented:
> I like the general idea (in fact I'm pretty sure I've got an old branch somewhere that does something similar but it's currently lost in the ether ;).
> 
> Also see this related issue: #251.

Viper certainly looks good overall although I'm not clear from reading the docs if it can merge configs read from several files like BurntSushi (sort of) can.

> I've added a few comments but I agree with @fidencio: this needs very careful documentation (and lots of tests). The way configuration is handled is already pretty complex so although this will allow more flexibility, we've got to be careful here.
> 
> See https://github.com/kata-containers/kata-containers/tree/main/src/runtime#runtime-configuration-file-location for the existing details on config priority.
> 
> As a minimum, I think we need new doc to answer the following sorts of questions to avoid confusion:

These are relevant questions, thanks! I'll try to answer them below. It's useful to keep in mind that most of the answers kind of follow from the fact that this PR basically just extends the existing `decodeConfig()` behaviour by additionally decoding and merging fragments that correspond to the config file the function called on.

>     * Can Kata run with only a set of fragments (aka no `configuration.toml` file)?

Regarding this PR, it cannot.  This follows from the fact that `decodeConfig()` will bail out before it gets to fragment handling if it cannot read the base config it was called on. I think this would be fairly easy to change though if desired.

>     * If a fragment sets a value that is set to a different value in the config, which one takes priority?

The last file to change a value takes priority.  Since fragments are decoded after the config, the fragment will take priority. 

>     * What happens if a fragment is invalid? Is it ignored, or does the runtime fail?

This is admittedly an underdeveloped aspect of the PR as I wasn't sure what the desired behaviour would be. I welcome a discussion to determine this and I'm ready to change the PR based on the outcome. Currently, an invalid fragment is basically ignored with the exception of an undecoded key error (essentially a typo in a key name) in which case the undecoded key(s) are logged but the rest of the fragment is used (I believe this is identical to how this is handled by the main config decoding).

>     * Can fragment files be empty?

An empty fragment file is a no-op.

>     * Can fragment files be sym-links?

This basically depends on how BurntSushi handles links I guess.  While I don't think the library documents this explicitly, in practice symlinks seem to be handled correctly.  I think it might be a good idea to call `ResolvePath()` on fragment paths to make the behaviour identical to the main config file handling.

>     * Can fragments be ordered to guarantee that one will be loaded before another?

The fragment directory (`config.d`) is enumerated by [ioutil.ReadDir](https://pkg.go.dev/io/ioutil#ReadDir) which is documented to sort the contents by filename.

>     * Considering `kata-runtime --show-default-config-paths`, currently the first configuration file found will be read and any others ignored. This PR potentially makes this logic less clear.

Hm, can you be more specific here?  I believe this PR doesn't affect this logic in any way (keep in mind that it's just an extension of `decodeConfig()`, anything that happens at higher layers should not be affected).

> For example, what happens in these scenarios?:
> `/etc/kata-containers/configuration.toml` 	`/etc/kata-containers/config.d/` 	`/usr/share/defaults/kata-containers/configuration.toml` 	`/usr/share/defaults/kata-containers/config.d/`
> empty 	does not exist / no files 	empty 	contains valid fragments
> exists 	does not exist / no files 	does not exist 	contains valid fragments
> exists 	contains invalid fragments 	does not exist 	contains valid fragments
> does not exist 	contains valid fragments 	does not exist 	contains valid fragments
> does not exist 	contains valid fragments 	exists 	contains valid fragments

The decision between `/etc/kata-containers/configuration.toml` and `/usr/share/defaults/kata-containers/configuration.toml` happens at a higher layer and is not affected by this PR.  Then, fragments are only decoded if _their corresponding_ `configuration.toml` exists (even if it's empty I believe).

> Finally, whatever we do here needs to be mirrored in the new `runtime-rs`, so let's get some thoughts from @bergwolf and @liubin.

I'm happy to work on reimplementing this in `runtime-rs`.

jodh-intel commented:
Thanks for the details @pmores but I think this needs to be documented for users (maybe in https://github.com/kata-containers/kata-containers/tree/main/src/runtime#runtime-configuration-file-location).

fidencio commented:
@pmores, I just tested your branch and it does work as expected, **very nice work on this one!**

As @jodh-intel mentioned, and I agree with him, we need to document it very well, but I think this is a very solid PR that only needs documentation tweaks plus make sure the tools we have behave accordingly with this new change.

A huge thumbs-up from me.

fidencio commented:
@pmores, in my mind a drop-in augments the config.  If there's something invalid in the fragment, it should make the whole configuration invalid.  Logging then which drop-in file brought in the invalid fragment is a very nice plus.

pmores commented:
@fidencio Yes, I'm also leaning towards that solution. I'll wait though at least for James' opinion before updating the PR.

pmores commented:
@fidencio @jodh-intel Alright, thanks!

A note about logging - the whole `config.go` file does very little explicit logging, it prefers to return (string) errors and I'm following suit to be consistent with the style of the surrounding code.  IIRC the only place where I log anything is for undecoded keys as I don't treat that as an error, again to be consistent with decoding of the main `configuration.toml`.

So any fatal problem while decoding a fragment file returns a string error which describes the problem and includes the name of the invalid file.  I hope that this fulfills your request - if not, please let me know, thanks!

jodh-intel commented on src/runtime/pkg/katautils/config.go:
```diff
@@ -1228,9 +1244,176 @@ func decodeConfig(configPath string) (tomlConfig, string, error) {
+	files, err := ioutil.ReadDir(dropInDir)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("error reading '%s' directory: %s\n", dropInDir, err)
```
Can you remove all the newlines in these errors and switch `'%s'` to `%q` (equivalent) here and elsewhere:

```suggestion
			return fmt.Errorf("error reading %q directory: %s", dropInDir, err)
```

pmores replied:
Sure, will do!

jodh-intel reviewed (commented):
Thanks @pmores - a few comments...

Also, it's not strictly required for this PR, but reading it made me wonder if we should warn/error if any config fragment is writable by "other" (`chmod +o`) as that would be good security practise and work work well with #3905 (which is in desparate need of reviews btw ;)

pmores commented on src/runtime/pkg/katautils/config.go:
```diff
@@ -166,6 +168,20 @@ type agent struct {
+	clone.Hypervisor = make(map[string]hypervisor)
+	clone.Agent = make(map[string]agent)
+
+	for key, value := range orig.Hypervisor {
+		clone.Hypervisor[key] = value
+	}
+	for key, value := range orig.Agent {
+		clone.Agent[key] = value
+	}
```
@jodh-intel The rest (the members of `tomlConfig` that are held by value) is handled by the line 172.  It's just the maps that need to be handled explicitly as the assignment on line 172 results in only a shallow copy of the `map` members.

BTW I tried to find a less verbose way to clone a `map` in Golang, like a standard library helper function, but a bit to my surprise, cloning by hand seems to be the idiomatic way to do that in Golang.  Admittedly I'm no Golang wizard though so if there is a better implementation of cloning a `tomlConfig` I'm happy to learn!

pmores commented on src/runtime/pkg/katautils/config.go:
```diff
@@ -1228,9 +1244,176 @@ func decodeConfig(configPath string) (tomlConfig, string, error) {
+// identified by the TOML tag in 'key' into the corresponding field in
+// 'targetConf'.
+func applyAgentKey(sourceConf tomlConfig, key []string, targetConf *tomlConfig) error {
+	agentName := key[0]
+	tomlKeyName := key[1]
```
The `len` check is performed by the caller (see line 1321 in `applyKey()`).  I moved the check there as it results in more compact code, ensures that whatever is called by `applyKey()` (now or in the future) is guaranteed to have the length already checked, and it's also a tiny optimisation as well (we exclude keys that don't require additional handling as early as possible).

jodh-intel replied:
I see that, but we try to ensure that all functions are coded defensively. Yes, the code works now, but the caller might be reworked in the future and silently break. Speaking of which, please add unit tests for all the new functions you've added on this PR.

pmores replied:
James, thanks for your comments!

As for testing: the only function that this PR adds is really just `decodeDropIns()`, the rest are just details of its implementation. I know this is not very obvious but I couldn't find a way within the context I was working in to make it more explicit. Ideally, this PR would be a package with `decodeDropIns()` becoming `DecodeDropIns()` and the rest  private but seeing as the whole configuration handling is in `katautils` (not even `config` or similar), I thought adding a package for my changes wouldn't be very fitting.

In my experience, testing private implementation details is generally not recommended as it makes for brittle tests that don't add much if anything at all as they wouldn't in this case - I believe that testing `decodeDropIns()` fully exercises the private implementation functions as well.  Of course I'm happy to add more tests of `decodeDropIns()` though if deemed necessary!

As for moving the `len` checks to the lower-level functions, I really don't think this would add to defensive coding.  `applyAgentKey()` and `applyHypervisorKey()` are just helpers of `applyKey()`, not API functions that could be called from anywhere. They would actually be quite hard to call correctly from anywhere but `applyKey()`, and it's not clear why anyone would even want to do that as their purpose is very narrow and tailored to what a single specific caller needs.

Apart from that, I believe the most likely future refactoring would be if another member of `tomlConfig` needs the special treatment `applyKey()` provides. In other words, if there's another member of `tomlConfig` which is a `map` - `runtime` comes to mind (?).  In that case, `applyXyzKey()` would need to be added and its implementation would benefit from being safe _by default_ from out-of-bounds key array accesses instead of having to keep in mind that it has to check bounds.

I'll make the requested changes if you insist, this is just to let you know why I wouldn't necessarily consider them beneficial.

pmores commented:
Pushed a new version which:
- implements the agreed-upon behaviour on invalid fragment file
- adds documentation
- implements @jodh-intel 's request for a formatting change of error strings.

pmores commented on src/runtime/README.md:
```diff
@@ -86,6 +86,26 @@ following locations (in order):
+Non-existent or empty `config.d` directory is not an error (in other words,
+not using configuration file fragments is fine).  On the other hand, if
+fragments are used, they must be valid - any errors while parsing fragments
+are treated the same as errors while parsing the main configuration file.  A
```
Yes, I can add some (malformed TOML or failure to open/read the files could be some of the more common ones and worth a mention).

The whole point here though is that the error handling is the same as for the main `configuration.toml`.  BTW this gave me a little trouble because I realised that the special case of undecoded keys (like typos in key names, or maybe obsoleted keys) are not considered an error by `configuration.toml` decoding when I believe they should be.  Fixing this might not be trivial though as it might break existing configs.

c3d commented:
> > @pmores, I'm thinking about users using this one and I have a question for you (and also for the other folks who'll review this one).
> > Consider I'm using kata-deploy, and while tweaking things here in there I connected to the node and added my `config.d` folder and dropped in my files there. What would be the expected kata-deploy behaviour from your POV? Shall it remove the `config.d` directory entirely? Shall the `config.d` directory be kept there?
> 
> Admittedly I'm not familiar enough with the behaviour kata-deploy aims for to give an authoritative answer, but my initial instinct would probably be to preserve site-specific customisations. (?)

I believe that the problem @fidencio is alluding to is that we may end with unpredictable configurations if there are things in there that you can't easily remove.

The problem is not new, it also happens with configuration changes under `/etc` not being removed by `kata-deploy`. So far, there is a sort of tacit understanding that if you care about your configuration being preserved, you put it under `/etc`, if you want your changes to be overwritten at the next update, you use the file under `/usr/share` or defaults or wherever your distro puts them.

With `config.d`, we add another layer to this cake, which is that even if `kata-deploy` updates everything in the default config file locations, it still may not erase deop-ins. Or alternatively, we could modify it to also remove the `config.d` there. I suspect @fidencio is subtly suggesting that you add this change to your series 😄 

> 
> Another way of thinking about this that comes to my mind would be based on realising that drop-ins are just an extension of the basic configuration in `configuration.toml`, so the answer should be the same as the answer to the question "what does kata-deploy do with local modifications to `configuration.toml`?". If it preserves them it should probably preserve drop-ins as well. If kata-deploy resets `configuration.toml` it should probably get rid of drop-ins as well.

Hence my conclusion above…

c3d commented on src/runtime/pkg/katautils/config.go:
```diff
@@ -1232,6 +1247,70 @@ func decodeConfig(configPath string) (tomlConfig, string, error) {
+	configDir := filepath.Dir(mainConfigPath)
+	dropInDir := filepath.Join(configDir, "config.d")
+
+	files, err := ioutil.ReadDir(dropInDir)
```
In the documentation and discussion, you stated several times that drop-in files would be read in "alphabetical order". This code uses `ReadDir`, but does not sort the output. The fact that `readdir` does not guarantee order is discussed here https://stackoverflow.com/questions/8977441/does-readdir-guarantee-an-order. Here is an example of `ReadDir` output not being in alphabetical order in Go. https://thedeveloperblog.com/go/readdir-go.

So I think that you probably want to add a sort here to ensure the kind of stable semantics we want.

pmores replied:
@c3d Thanks for your review!  Well, I'm not quite convinced we have a problem here since the former link pertains to POSIX `readdir()` which doesn't sort the output but is not the function called here, and the latter one pertains to [Readdir() in os.File](https://pkg.go.dev/os#File.Readdir) which also doesn't sort the output but, again, is not the function called here either.  What's called here is [ioutil.ReadDir()](https://pkg.go.dev/io/ioutil#ReadDir) which is documented to sort the output by filename.

Now, it's true that the docs don't specify "alphabetical order" explicitly so you're right, I might have assumed more than I was afforded by the docs. I'd say that it's a safe assumption though, but if there's a consensus among reviewers that we should sort as well this should be easy to add.

pmores reviewed (commented):
Damn, I apparently pushed the wrong button when submitting a comment and now I'm reviewing my own stuff! :-( ;-D  Trying to get rid of this by submitting general feedback (this message) without approval...

pmores commented:
> Assuming the drop in cfg files are for users to customize existing configs, can we simplify a bit and only handle `/etc/kata-containers/config.d/` and leave `/usr/share/defaults/kata-containers/config.d/` out of math? We are not going to ship any default drop in cfg files, are we?

Hi @bergwolf , this PR is oblivious of the actual configuration locations. So if we don't want to distribute drop-in cfg files in `/usr/share/defaults/kata-containers/config.d/` (and I agree it does't seem to make much sense to do so) we just don't create that directory and everything will continue to work like it has so far, no code changes in this PR needed.

pmores commented:
Pushed requested documentation fixes.

Thanks @c3d for your review!

liubin commented:
@pmores The CI complains about a not found URL for [kata-log-parser](https://github.com/kata-containers/tests/tree/main/cmd/log-parser), you need to do a rebase on the latest main branch.

pmores commented:
@liubin Done, thanks for pointing this out! Awaiting CI's verdict. ;-)

fidencio commented:
/test

fidencio commented:
@pmores, could you rebase so we hopefully can get rid of the static checks errors?

pmores commented:
@fidencio Done!

fidencio commented:
/test

pmores commented:
Static check failed due to "parseable" in `README.md`. Changed to "parsable" and resubmitted. Hopefully this will pass.

fidencio commented:
@pmores, you'll have to add the word as part of the dictionary, in the tests repo :-/
```
[kata-spell-check.sh:275] WARNING: Word 'parsable': did you mean one of the following?: parable, parables, passable, par sable, par-sable, pars able, pars-able, palpable
ERROR: Spell check failed for file: 'src/runtime/README.md'
```

pmores commented:
Not even "parsable" passes, reworded.

pmores commented:
/test

gkurz commented:
/retest-arm

fidencio commented:
/retest-arm

fidencio commented:
> `17:36:04 --- FAIL: TestIoCopy (20.09s) 17:36:04 testing.go:1152: race detected during execution of test `

This is known to be racy.  Not sure about the other one. cc @cmaf.

fidencio reviewed (approved):
lgtm, thanks @pmores!
